sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -881,6 +881,16 @@ object SQLConf {
       .intConf
       .createWithDefault(10000)
 
+  val HIVE_STYLE_PARTITION_OVERWRITE =
+    buildConf("spark.sql.hiveStylePartitionOverwrite")
+      .doc("When insert overwrite a partitioned table with dynamic partition columns, Spark " +
+        "will follow Hive's behavior if this config is enabled, i.e., only delete partition " +
+        "directories which have data written into it. By default this config is disabled " +
+        "to keep the previous behavior, which means Spark will delete all partition directories " +
+        "that match the static partition values provided in the insert statement.")
+      .booleanConf
+      .createWithDefault(false)
```

Comment:
a lot of tests will fail because we explicitly assert the old behavior, but I can try